### PR TITLE
[Navigation] Add a new prop to truncate nav item text

### PR DESCRIPTION
**Pull request by Rmnlly**
### WHY are these changes introduced?
Overwrites have been used in admin to get navigation item text to truncate with ellipsis since the app pinning project. This is an effort to port the changes to Polaris to reduce the overwriting and tech debt in core. 

Fixes https://github.com/Shopify/core-workflows/issues/561 

### WHAT is this pull request doing?
I've added a prop to allow text truncation in navigation items. It applies a class that stops the line wrapping of item text and instead does an ellipsis truncation through CSS.

Screen Shot 2022-11-04 at 12 20 29 PM

There is a new story for the truncation enabled items. Here you can see anything going beyond the width of the text container will be truncated with ellipsis. Notice that "Inventory" is fine as it not long enough to trigger it. Secondary navigation items are also treated the same as their parent. 

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
    </Page>
  );
}
```

</details>

### 🎩 checklist

- [x] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [x] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

**Comment by alex-page** [comment]
@Rmnlly should nav items just be truncated by default? I wonder how this could impact translations... My gut feeling is we shouldn't add a property here and items should just truncate if necessary.

**Comment by Rmnlly** [comment]
@alex-page (Sorry I messed up the reply haha)

Do you mean instead of a prop to allow this functionality, instead have this be the default for all navigation items? I imagined there are cases where people would prefer to have the multiline full title displayed. As for translations I'm not sure what the impact might be besides the ellipsis having an alternative meaning in another language?

**Comment by alex-page** [comment]
In my opinion we should have a strong pattern here that works across all items. Either they truncate or don't. I personally prefer the items to go onto two lines as for translations this can easily happen.

Do you have a strong example for why we would make truncation an option?

**Comment by kyledurand** [comment]
> Do you have a strong example for why we would make truncation an option?

It's only for apps according to a slack convo but that doesn't mean it couldn't get misused. I'd also prefer to never truncate. If we absolutely need to because of ridiculously long app names then maybe line-clamp with truncate is an option? Looks like we'd be able to use `-webkit-line-clamp` in the admin now: https://caniuse.com/mdn-css_properties_-webkit-line-clamp

**Comment by Rmnlly** [comment]
@kyledurand 

I hadn't heard of line-clamp but it actually solved a bug I was dealing with so thats awesome! I updated the badge spacing too.

File: polaris-react/src/components/Navigation/Navigation.scss [hunk 1]
```diff
@@ -92,6 +92,14 @@ $disabled-fade: 0.6;
 .Item {
   @include nav-item-attributes;
   position: relative;
+
+  &:first-child {
+    margin-left: var(--p-space-2);
```
Comment (kyledurand) [inline polaris-react/src/components/Navigation/Navigation.scss]:
```suggestion
    margin-inline-start: var(--p-space-2);
```

File: polaris-react/src/components/Navigation/Navigation.scss [hunk 2]
```diff
@@ -92,6 +92,14 @@ $disabled-fade: 0.6;
...
+  }
+
+  &:last-child {
+    margin-right: var(--p-space-2);
```
Comment (kyledurand) [inline polaris-react/src/components/Navigation/Navigation.scss]:
```suggestion
    margin-inline-end: var(--p-space-2);
```

File: polaris-react/src/components/Navigation/_variables.scss [hunk 3]
```diff
@@ -28,7 +28,7 @@ $nav-animation-variables: (
   align-items: flex-start;
   max-width: 100%;
   padding: 0 var(--p-space-1) 0 var(--p-space-3);
-  margin: 0 var(--p-space-2);
+  margin: 0;
```
Comment (kyledurand) [inline polaris-react/src/components/Navigation/_variables.scss:31]: Why was this moved to margin left / right, first / last child?

File: .changeset/tidy-waves-live.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+'@shopify/polaris': minor
+---
+
+A new prop has been added (truncateText) to allow navigation item text to be truncated instead of flowing onto multiple lines
```
Comment (kyledurand) [inline .changeset/tidy-waves-live.md]:
```suggestion
Added `truncateText` prop to `Navigation.Item` to prevent text wrapping
```

File: polaris-react/src/components/Navigation/Navigation.scss [hunk 5]
```diff
@@ -213,13 +221,22 @@ $disabled-fade: 0.6;
   }
 }
 
+.Text-truncated {
+  // stylelint-disable-next-line value-no-vendor-prefix
+  display: -webkit-box;
+  overflow: hidden;
+  -webkit-line-clamp: 1;
```
Comment (kyledurand) [inline polaris-react/src/components/Navigation/Navigation.scss:226]: I was suggesting a global line clamp rule instead of using it here with a value of one. Does `@include truncate;` not work here?

**Review by kyledurand — COMMENTED** [review]
I don't love that we need to add a new prop for this and truncating text in the nav. My original suggestion was to line clamp to two lines and then truncate for all nav items. This seems like it could be a better middle ground but might require some UX buy in.

Things that make me nervous about this:
1. Could be misused and create a worse merchant experience
2. Moving the margins to first-child / last-child. Why was that necessary?
3. Using line clamp for single line truncation seems like overkill. Can you change that to use the truncate mixin?
4. Can you use `/snapit` and test this in web? Share the spin link when you can

File: polaris-react/src/components/Navigation/_variables.scss (hunk 3 above)
Comment (Rmnlly) [reply polaris-react/src/components/Navigation/_variables.scss:31]: So this was moved because originally the nav-item used this margin to fit into the nav, but it was also using this margin between itself and secondary actions. If you leave this on the nav-item then the secondary action is pushed away and the truncation leaves little space left for the app titles. So having this on first/last makes it so when we have the secondary action it'll get rid of its right margin and when we don't it fits how it would originally.

File: polaris-react/src/components/Navigation/Navigation.scss (hunk 5 above)
Comment (Rmnlly) [reply polaris-react/src/components/Navigation/Navigation.scss:226]: I wasn't aware of the `@include truncate;` mixin, but after trying it I run into the issue I had originally. So when using the `overflow: hidden;` + `text-overflow: ellipsis;` method (that the @truncate mixin uses) you need to set a specific width on the element being truncated. This becomes an issue because the secondary action needs to be accounted for AND the badge needs to be accounted for, so in my original PR I had that, but if you have both a badge and secondary action it becomes messy and doesn't quite fit. 

The ` -webkit-line-clamp: 1;` method doesn't require a specified width on the element which is convenient because with that we can fit both badges and secondary actions (which does happen in the admin nav).

**Review by kyledurand — APPROVED** [review]

**Comment by Rmnlly** [comment]
@alex-page What do you think after the changes/the extra context added? Just making sure you're okay if we're going to move forward since I added ya originally.

**Review by alex-page — APPROVED** [review]

File: polaris-react/src/components/Navigation/Navigation.scss [hunk 6]
```diff
@@ -92,6 +92,14 @@ $disabled-fade: 0.6;
...
   @include nav-item-attributes;
   position: relative;
+
+  &:first-child {
+    margin-inline-start: var(--p-space-2);
+  }
+
+  &:last-child {
+    margin-inline-end: var(--p-space-2);
+  }
```
Comment (henryyi) [inline polaris-react/src/components/Navigation/Navigation.scss]: I don't think there is a case where `Item` isn't a first child.
In the case of the end-margin what we're really accomplishing here is removing the extra space when there is a secondary item. So perhaps it can be phrase like so:
```sass
margin-inline-start: var(--p-space-2);

&:only-child {
    margin-inline-end: var(--p-space-2);
  }
```

File: polaris-react/src/components/Navigation/Navigation.scss [hunk 7]
```diff
@@ -213,13 +221,22 @@ $disabled-fade: 0.6;
   }
 }
 
+.Text-truncated {
+  // stylelint-disable-next-line value-no-vendor-prefix
+  display: -webkit-box;
+  overflow: hidden;
+  -webkit-line-clamp: 1;
+  -webkit-box-orient: vertical;
+  word-break: break-all;
```
Comment (henryyi) [inline polaris-react/src/components/Navigation/Navigation.scss:228]: is `word-break` necessary if it's just one line?

Comment (henryyi) [reply polaris-react/src/components/Navigation/Navigation.scss:228]: guess it makes a difference for larger words

**Review by henryyi — APPROVED** [review]
🎩 and changes look good. Only minor non-blocking comment.

File: polaris-react/src/components/Navigation/Navigation.scss (hunk 7 above)
Comment (Rmnlly) [reply polaris-react/src/components/Navigation/Navigation.scss:228]: Yeah it breaks better for longer sentences from when I was testing!

File: polaris-react/src/components/Navigation/Navigation.scss (hunk 6 above)
Comment (Rmnlly) [reply polaris-react/src/components/Navigation/Navigation.scss]: For sure, I had it the other way originally to showcase symmetry but this is good, Implemented!
